src/pandapipes/pf/derivative_calculation.py

```diff
@@ -94,29 +110,121 @@ def calculate_derivatives_thermal(net, branch_pit, node_pit, _):
-    cp_n = fluid.get_heat_capacity(t_init_n)
+    t_init_n = node_pit[:, TINIT_NODE]
+    cp_i = fluid.get_heat_capacity(t_init_i)
     cp_i1 = fluid.get_heat_capacity(t_init_i1)
+    cp_n = fluid.get_heat_capacity(t_init_n)
     t_amb = branch_pit[:, TEXT]
     length = branch_pit[:, LENGTH]
     alpha = branch_pit[:, ALPHA] * np.pi * branch_pit[:, D]
     tl = branch_pit[:, TL]
     qext = branch_pit[:, QEXT]
-    t_m = (t_init_i1 + t_init_i) / 2
-    m_m = (m_init_i + m_init_i1) / 2
     no_cp = branch_pit[:, BRANCH_TYPE] != CIRC
     infeed_node = np.setdiff1d(from_nodes[no_cp], to_nodes[no_cp])
 
-    branch_pit[:, JAC_DERIV_DT] = - cp * m_m + alpha / 2 * length
-    branch_pit[:, JAC_DERIV_DTOUT] = cp * m_m + alpha / 2 * length
-    branch_pit[:, LOAD_VEC_BRANCHES_T] = cp * m_m * (-t_init_i + t_init_i1 - tl) - alpha * (
-                t_amb - t_m) * length + qext
+    node_pit[:, INFEED] = False
+    node_pit[infeed_node, INFEED] = True
 
-    branch_pit[:, JAC_DERIV_DT_NODE] = - m_init_i * cp_n
+    node_pit[:, LOAD_T] = node_pit[:, LOAD] * cp_n * t_init_n + node_pit[:, MDOTSLACKINIT] * cp_n * t_init_n
+    node_pit[:, JAC_DERIV_DT_LOAD] = - node_pit[:, LOAD] * cp_n
+    node_pit[:, JAC_DERIV_DT_SLACK] = - node_pit[:, MDOTSLACKINIT] * cp_n
+
+    branch_pit[:, JAC_DERIV_DT_NODE] = - m_init_i * cp_i
     branch_pit[:, JAC_DERIV_DTOUT_NODE] = m_init_i1 * cp_i1
-    branch_pit[:, LOAD_VEC_NODES_FROM_T] = m_init_i1 * t_init_n * cp_n
+    branch_pit[:, LOAD_VEC_NODES_FROM_T] = m_init_i1 * t_init_i * cp_i
     branch_pit[:, LOAD_VEC_NODES_TO_T] = m_init_i1 * t_init_i1 * cp_i1
 
+    if get_net_option(net, "transient"):
+        rho = get_branch_real_density(fluid, node_pit, branch_pit)
+        area = branch_pit[:, AREA]
+        tvor = branch_pit[:, T_OUT_OLD]
+        delta_t = get_net_option(net, "dt")
+
+        branch_pit[:, LOAD_VEC_BRANCHES_T] = (
+                rho * area * cp * (t_init_i1 - tvor) * (1 / delta_t) * length
+                + cp * m_init_i * (-t_init_i + t_init_i1 - tl)
+                - alpha * (t_amb - t_init_i1) * length + qext
+        )
+
+        branch_pit[:, JAC_DERIV_DT] = - cp * m_init_i
+        branch_pit[:, JAC_DERIV_DTOUT] = rho * area * cp / delta_t * length + cp * m_init_i + alpha
+
+        branches_active_ht = get_lookup(net, "branch", "active_heat_transfer")
+        branches_zero_fl = get_lookup(net, "branch", "zero_flow")[branches_active_ht]
+        if np.any(branches_zero_fl):
+            # TODO: maybe replace this statement with a component lookup
+            zero_length = np.isclose(branch_pit[:, LENGTH], 0, rtol=1e-6, atol=1e-10)
+            mask = zero_length & branches_zero_fl
+            if np.any(mask):
+                branch_pit[mask, LOAD_VEC_BRANCHES_T] = (
+                        rho[mask] * area[mask] * cp[mask] * (t_init_i1[mask] - tvor[mask]) * (1 / delta_t)
+                        - alpha[mask] * (t_amb[mask] - t_init_i1[mask])
+                )
```

Comment:
Here the cp*m_init_i term can be dropped because the flow here is zero? 

Why can the length be removed?

Reply:
> For Understanding: Here the cp*m_init_i term can be dropped because the flow here is zero?

Yes

> Why can the length be removed?

We mainly need to solve an equation for a temperature that is not quite relevant, as without flow the outlet temperature has no impact on the downstream node equation. With length being 0, this part becomes annoying, since this fact leads to all-zero rows in the system matrix and thus a matrix rank problem.